### ISSUE TITLE
HLint suggests to remove -XTypeApplications when there is a type-level type application

### ISSUE THREAD

**Issue by goldfirere** [ndmitchell/hlint#1334]
If I say

```hs
{-# LANGUAGE TypeApplications, DataKinds #-}

module Bug where

import Type.Reflection

x :: Typeable b => TypeRep @Bool b
x = typeRep
```

then HLint says I should drop the `TypeApplications`, even though I need it for the `@Bool`.

Seems like HLint should probably look for type applications in types, too.

**Comment by shayne-fletcher** [comment]
@goldfirere I'm not sure that you do need `TypeApplications` here? It seems the module requires `DataKinds` but removing `TypeApplications` doesn't appear to an effect.

**Comment by shayne-fletcher** [comment]
> @goldfirere I'm not sure that you do need `TypeApplications` here? It seems the module requires `DataKinds` but removing `TypeApplications` doesn't appear to an effect.

Never mind. It would seem that `TypeApplications` is enabled by default. Disabling it explicitly causes the anticipated compiler error.

**Comment by goldfirere** [comment]
Also, I was working in GHC 8.10.7, when `-XTypeApplications` is not on by default, though I don't know how HLint would know this.

**Comment by ndmitchell** [comment]
My guess is indeed that HLint is forgetting to look for type applications in types. The line that checks for the various conditions is https://github.com/ndmitchell/hlint/blob/master/src/Hint/Extensions.hs#L378 - `isTypeApp` is defined at https://hackage.haskell.org/package/ghc-lib-parser-ex-9.2.0.1/docs/Language-Haskell-GhclibParserEx-GHC-Hs-Expr.html#v:isTypeApp. Given it takes an Expr, I guess we need another variant that looks for TypeApp in types.

**Comment by ndmitchell** [comment]
I see @shayne-fletcher beat me to it! Will be fixed once we have a GHC 9.2.2 or decide to backport the relevant fixes that are blocking HLint.
